flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTimeoutTest.java

```diff
@@ -0,0 +1,200 @@
+        TestProcessingTimeService tpts = context.getTestProcessingTimeService();
+        TimeoutWriter writer = new TimeoutWriter(context, 1, 10, 100, false);
+        writer.setShouldFailRequest(true);
+        tpts.setCurrentTime(0);
+        writer.write("1", null);
+        tpts.setCurrentTime(110);
+        // deliver initial message after timeout, message request would fail but elements should be
+        // discarded as timeout already occurred
+        writer.deliverMessage();
+        // flush outstanding mailbox messages to resubmit timed out elements
+        writer.flush(false);
+        assertThat(destination).containsExactly(1L);
+    }
+
+    @Test
+    void writerShouldFailOnFatalError() throws Exception {
+        TestSinkInitContextAnyThreadMailbox context = new TestSinkInitContextAnyThreadMailbox();
+        TestProcessingTimeService tpts = context.getTestProcessingTimeService();
+        TimeoutWriter writer = new TimeoutWriter(context, 1, 10, 100, true);
+        tpts.setCurrentTime(0);
+        writer.setFatalError(new FlinkRuntimeException("Fatal error"));
+        writer.write("1", null);
+        writer.deliverMessage();
+        assertThatExceptionOfType(FlinkRuntimeException.class)
+                .isThrownBy(() -> writer.flush(false))
+                .withMessage("Fatal error");
+    }
+
+    /** Test that the writer can handle a timeout. */
+    private class TimeoutWriter extends AsyncSinkWriter<String, Long> {
+        private Exception fatalError;
+        private final CountDownLatch completionLatch;
+        private Thread submitThread;
+
+        private boolean shouldFailRequest = false;
+
+        public TimeoutWriter(
+                WriterInitContext writerInitContext,
+                int maxBatchSize,
+                long maximumTimeInBufferMs,
+                long requestTimeout,
+                boolean failOnTimeout) {
+            super(
+                    (ElementConverter<String, Long>) (element, context) -> Long.parseLong(element),
+                    writerInitContext,
+                    AsyncSinkWriterConfiguration.builder()
+                            .setMaxBatchSize(maxBatchSize)
+                            .setMaxBatchSizeInBytes(Long.MAX_VALUE)
+                            .setMaxInFlightRequests(Integer.MAX_VALUE)
+                            .setMaxBufferedRequests(Integer.MAX_VALUE)
+                            .setMaxTimeInBufferMS(maximumTimeInBufferMs)
+                            .setMaxRecordSizeInBytes(Long.MAX_VALUE)
+                            .setRequestTimeoutMS(requestTimeout)
+                            .setFailOnTimeout(failOnTimeout)
+                            .build(),
+                    Collections.emptyList());
+            this.completionLatch = new CountDownLatch(1);
+        }
+
+        @Override
+        protected void submitRequestEntries(
+                List<Long> requestEntries, ResultHandler<Long> resultHandler) {
+            submitThread =
```

Comment:
Maybe use threadpool, instead of creating and destroying threads for each test?

Reply:
good suggestion